main.py

```diff
@@ -83,34 +92,171 @@ def run(driver, shoe_type, username, password, url, shoe_size, login_time=None,
+                    break
+                except Exception as e:
+                    LOGGER.exception("Failed visibility check for Payment: " + str(e))
+                
+                try:
+                    xpath = SUBMIT_BUTTON_XPATH
+                    check_submit_button(driver=driver, xpath_o=xpath)
+                    LOGGER.info("Submit Button appeared!")
+                    skip_add_address = True
+                    skip_select_shipping = True
+                    skip_payment = True
+                    break
+                except Exception as e:
+                    LOGGER.exception("Failed visibility check for Submit Button: " + str(e))
+                    
+                if checkout_num_retries_attempted < checkout_num_retries:
+                    checkout_num_retries_attempted += 1
+                    continue
+                else:
+                    LOGGER.info("Too many iterations through checkout element check. Terminating check...")
+                    break
+                    
+            if skip_add_address is False and shipping_address:
+                try:
+                    click_add_new_address_button(driver=driver)
+                except Exception as e:
+                    LOGGER.exception("Failed to click Add New Address button: " + str(e))
+                    six.reraise(Exception, e, sys.exc_info()[2])
+                              
+                try:
+                    shipping_address_data=json.loads(shipping_address)
+                    input_address(driver=driver, shipping_address=shipping_address_data)
+                except Exception as e:
+                    LOGGER.exception("Failed to select address: " + str(e))
+                    six.reraise(Exception, e, sys.exc_info()[2])
+            
+                try:
+                    click_save_button(driver=driver)
+                except Exception as e:
+                    LOGGER.exception("Failed to click save button: " + str(e))
+                    six.reraise(Exception, e, sys.exc_info()[2])
+            
+            if skip_select_shipping is False:
+                try:
+                    select_shipping_option(driver=driver, shipping_option=shipping_option)                    
+                except Exception as e:
+                    LOGGER.exception("Failed to select shipping option: " + str(e))
                     six.reraise(Exception, e, sys.exc_info()[2])
 
                 try:
                     click_save_button(driver=driver)
+                except StaleElementReferenceException:
+                    LOGGER.info("Failed to click save button (StaleElementReferenceException). Retrying once...")
+                    retry_save = True
+                except Exception as e:
+                    LOGGER.exception("Failed to click save button: " + str(e))
+                    six.reraise(Exception, e, sys.exc_info()[2])
+                    
+                if retry_save:
```

Comment:
I think you can just get rid of `retry_save` entirely and do this logic where you do `retry_save = True`.